src/AzureAuth/Commands/Ado/CommandPat.cs

```diff
@@ -60,19 +111,161 @@ private enum OutputMode
 
+        [Option(ModeOption, ModeHelp, CommandOptionType.MultipleValue)]
+        private IEnumerable<AuthMode> AuthModes { get; set; } = new[] { AuthMode.Default };
+
+        [Option(DomainOption, DomainHelp, CommandOptionType.SingleValue)]
+        private string Domain { get; set; } = AzureAuth.Ado.Constants.PreferredDomain;
+
         /// <summary>
         /// Executes the command and returns a status code indicating the success or failure of the execution.
         /// </summary>
         /// <param name="logger">The <see cref="ILogger{T}"/> instance that is used for logging.</param>
+        /// <param name="publicClientAuth">An <see cref="IPublicClientAuth"/>.</param>
         /// <param name="eventData">Lasso injected command event data.</param>
         /// <returns>An integer status code. 0 for success and non-zero for failure.</returns>
-        public int OnExecute(ILogger<CommandPat> logger, CommandExecuteEventData eventData)
+        public int OnExecute(ILogger<CommandPat> logger, IPublicClientAuth publicClientAuth, CommandExecuteEventData eventData)
         {
-            logger.LogInformation("coming soon");
+            if (!this.ValidOptions(logger))
+            {
+                return 1;
+            }
+
+            var accessToken = this.AccessToken(publicClientAuth, eventData);
+            if (accessToken == null)
+            {
+                logger.LogError("Failed to acquire an Azure DevOps access token. Re-run with '--verbosity debug' for more info.");
+                return 1;
+            }
+
+            IPatCache cache = null;
+            try
+            {
+                cache = this.Cache();
+            }
+            catch (MsalCachePersistenceException e)
+            {
+                logger.LogError($"Failed to validate cache persistence: {e.Message}");
+                return 1;
+            }
+
+            var client = this.Client(accessToken.Token);
+            var manager = new PatManager(cache, client);
+
+            using (new CrossPlatLock(LockfilePath))
+            {
+                var pat = manager.GetPatAsync(this.PatOptions()).Result;
+
+                // Do not use logger to avoid printing PATs into log files.
+                Console.Write(FormatPat(pat, this.Output));
+            }
+
             return 0;
         }
+
+        private static string FormatPat(PatToken pat, OutputMode output) => output switch
+        {
+            OutputMode.None => string.Empty,
+            OutputMode.Status => $"\"{pat.DisplayName}\" valid until {pat.ValidTo:O}\n",
+            OutputMode.Token => $"{pat.Token}\n",
+            OutputMode.Base64 => $"{pat.Token.Base64()}\n",
+            OutputMode.Header => $"{pat.Token.AsHeader(AzureAuth.Ado.Authorization.Basic)}\n",
+            OutputMode.HeaderValue => $"{pat.Token.AsHeaderValue(AzureAuth.Ado.Authorization.Basic)}\n",
+            OutputMode.Json => $"{pat.AsJson()}\n",

```

Comment:
These could avoid potential usage issues by not adding newlines because the intent is for use by other programs, and forgetting to trim your strings is an easy pitfall.
```suggestion
            OutputMode.Token => $"{pat.Token}",
            OutputMode.Base64 => $"{pat.Token.Base64()}",
            OutputMode.Header => $"{pat.Token.AsHeader(AzureAuth.Ado.Authorization.Basic)}",
            OutputMode.HeaderValue => $"{pat.Token.AsHeaderValue(AzureAuth.Ado.Authorization.Basic)}",
            OutputMode.Json => $"{pat.AsJson()}",
```